### PR TITLE
FOCUS #366: SkuId, SkuPriceId, and Pricing-related columns review due to ChargeCategory and ChargeClass cleanup

### PR DESCRIPTION
**Pull request by ijurica**
ChargeCategory and ChargeClass cleanup impacted multiple columns. Based on current ChargeCategory references, the following columns must be revisited:

- SkuId
- SkuPriceId
- ContractedUnitPrice
- ListUnitPrice
- Pricingcategory
- PricingUnit
- PricingQuantity
- UsageQuantity
- UsageUnit
- ListCost
- ContractedCost
- EffectiveCost

### CODE REVIEW

File: supporting_content/columns/skupriceid.md [hunk 1]
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
...
+* Potato / Tomato v2 discussion:\
 <https://docs.google.com/document/d/18eL6G8WhbmEIHtrjqQlWqckgMRUQSs1aZwmwuRKQfqU/edit#heading=h.swm58hl317f3>
+
+### Impacts of 1.0 ChargeCategory and ChargeClass cleanup
+
+The following table serves as the basis for reviewing the SkuPriceId spec, as well as price, cost, quantity metrics, etc., impacted by the ChargeCategory and ChargeClass columns cleanup
+
+| ChargeCategory | ChargeClass | perSku/bulk                       | SkuId            | SkuPriceId       |
+|----------------|-------------|-----------------------------------|------------------|------------------|
+| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```
Comment (ljadvey) [inline supporting_content/columns/skupriceid.md:43]:
```suggestion
| Usage          | Original     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```

File: supporting_content/columns/skupriceid.md [hunk 2]
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
...
+| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Purchase       | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Adjustment     | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
+| Adjustment     | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Tax            | Regular     | MUST be bulk                      | MUST be null     | MUST be null     |
```
Comment (ljadvey) [inline supporting_content/columns/skupriceid.md:51]:
```suggestion
| Tax            | Original     | MUST be bulk                      | MUST be null     | MUST be null     |
```

File: supporting_content/columns/skupriceid.md [hunk 3]
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
...
+
+| ChargeCategory | ChargeClass | perSku/bulk                       | SkuId            | SkuPriceId       |
+|----------------|-------------|-----------------------------------|------------------|------------------|
+| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Purchase       | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Adjustment     | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
```
Comment (ljadvey) [inline supporting_content/columns/skupriceid.md:49]:
```suggestion
| Adjustment     | Original     | MAY be bulk                       | MAY be null      | MAY be null      |
```

File: supporting_content/columns/skupriceid.md [hunk 4]
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
...
+
+The following table serves as the basis for reviewing the SkuPriceId spec, as well as price, cost, quantity metrics, etc., impacted by the ChargeCategory and ChargeClass columns cleanup
+
+| ChargeCategory | ChargeClass | perSku/bulk                       | SkuId            | SkuPriceId       |
+|----------------|-------------|-----------------------------------|------------------|------------------|
+| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Purchase       | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
```
Comment (ljadvey) [inline supporting_content/columns/skupriceid.md:47]:
```suggestion
| Credit         | Original     | MAY be bulk                       | MAY be null      | MAY be null      |
```

File: supporting_content/columns/skupriceid.md [hunk 5]
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
...
+
+### Impacts of 1.0 ChargeCategory and ChargeClass cleanup
+
+The following table serves as the basis for reviewing the SkuPriceId spec, as well as price, cost, quantity metrics, etc., impacted by the ChargeCategory and ChargeClass columns cleanup
+
+| ChargeCategory | ChargeClass | perSku/bulk                       | SkuId            | SkuPriceId       |
+|----------------|-------------|-----------------------------------|------------------|------------------|
+| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```
Comment (ljadvey) [inline supporting_content/columns/skupriceid.md:45]:
```suggestion
| Purchase       | Original     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```

File: supporting_content/columns/skupriceid.md (hunk 1 above)
Comment (ljadvey) [reply supporting_content/columns/skupriceid.md:43]: Original change made on feedback from TF1 call

**Comment by ijurica** [comment]
### Changes overview

#### Adjusting normative paragraphs of multiple columns

The following table served as the basis for adjusting normative paragraphs of SkuPriceId spec, as well as price, cost, quantity metrics, etc., impacted by the ChargeCategory and ChargeClass columns cleanup

| ChargeCategory | ChargeClass | perSku/bulk                       | SkuId            | SkuPriceId       |
|----------------|-------------|-----------------------------------|------------------|------------------|
| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
| Purchase       | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Credit         | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
| Credit         | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Adjustment     | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
| Adjustment     | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Tax            | Regular     | MUST be bulk                      | MUST be null     | MUST be null     |
| Tax            | Correction  | MUST be bulk                      | MUST be null     | MUST be null     |

Same wording used in multiple columns:

- skuid.md
- skupriceid.md
- contractedunitprice.md
- listunitprice.md
- pricingcategory.md
- pricingquantity.md
- pricingunit.md
- usagequantity.md
- usageunit.md

**Example: PricingCategory**

> - PricingCategory MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.

#### Covering independent cost and/or pricing corrections

Note: Same wording used in multiple columns:

- contractedcost.md
- contractedunitprice.md
- listcost.md
- listunitprice.md
- pricingquantity.md

**Example: PricingQuantity**

> When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric, **except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently**.

**Review by ahullah — APPROVED** [review]

**Comment by udam-f2** [comment]
The group approved the change except for a minor update to lowercase the final 'MAY' -> may:
When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric, except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.

Also needs an update in changelog.

File: specification/columns/contractedunitprice.md [hunk 6]
```diff
@@ -2,7 +2,7 @@
...
+The ContractedUnitPrice column MUST be present in the billing data when the provider supports negotiated pricing concept. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When ContractedUnitPrice is present and not null, multiplying ContractedUnitPrice by [PricingQuantity](#pricingquantity) MUST equal [ContractedCost](#contractedcost), except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```
Comment (flanakin) [inline specification/columns/contractedunitprice.md:5]:
```suggestion
The ContractedUnitPrice column MUST be present in the billing data when the provider supports negotiated pricing concept. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When ContractedUnitPrice is present and not null, multiplying ContractedUnitPrice by [PricingQuantity](#pricingquantity) MUST equal [ContractedCost](#contractedcost), except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```

File: specification/columns/listcost.md [hunk 7]
```diff
@@ -2,7 +2,7 @@
...
 
-The ListCost column MUST be present in the billing data and MUST NOT be null. This column MUST be of type Decimal, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When [ListUnitPrice](#listunitprice) is present and not null, multiplying the ListUnitPrice by PricingQuantity MUST produce the ListCost.
+The ListCost column MUST be present in the billing data and MUST NOT be null. This column MUST be of type Decimal, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When [ListUnitPrice](#listunitprice) is present and not null, multiplying the ListUnitPrice by PricingQuantity MUST produce the ListCost, except in cases of [ChargeClass](#chargeclass) "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```
Comment (flanakin) [inline specification/columns/listcost.md:5]:
```suggestion
The ListCost column MUST be present in the billing data and MUST NOT be null. This column MUST be of type Decimal, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When [ListUnitPrice](#listunitprice) is present and not null, multiplying the ListUnitPrice by PricingQuantity MUST produce the ListCost, except in cases of [ChargeClass](#chargeclass) "Correction", which MAY address PricingQuantity or any cost discrepancies independently.
```

File: specification/columns/listunitprice.md [hunk 8]
```diff
@@ -2,7 +2,7 @@
...
+The ListUnitPrice column MUST be present in the billing data when the provider publishes unit prices exclusive of discounts. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When ListUnitPrice is present and is not null, multiplying ListUnitPrice by [PricingQuantity](#pricingquantity) MUST equal [ListCost](#listcost), except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```
Comment (flanakin) [inline specification/columns/listunitprice.md:5]:
```suggestion
The ListUnitPrice column MUST be present in the billing data when the provider publishes unit prices exclusive of discounts. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When ListUnitPrice is present and is not null, multiplying ListUnitPrice by [PricingQuantity](#pricingquantity) MUST equal [ListCost](#listcost), except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```

File: specification/columns/pricingcategory.md [hunk 9]
```diff
@@ -5,7 +5,7 @@ Pricing Category describes the pricing model used for a charge at the time of us
 The PricingCategory column adheres to the following requirements:
 
 * PricingCategory MUST be present in the billing data when the provider supports more than one pricing category across all SKUs and MUST be of type String.
-* PricingCategory MUST NOT be null if [ChargeCategory](#chargecategory) is "Usage" or "Purchase" and MUST be null for other ChargeCategory values.
+* PricingCategory MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.
```
Comment (flanakin) [inline specification/columns/pricingcategory.md:8]:
```suggestion
* PricingCategory MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.
```

**Comment by jpradocueva** [comment]
Approved by the group on April 25, with changes. MAY to may and add changes to changelog
> PricingQuantity MUST produce the ListCost, except in cases of [ChargeClass](#chargeclass) "Correction", which `MAY` address PricingQuantity and/or any cost discrepancies independently.

File: specification/columns/pricingquantity.md [hunk 10]
```diff
@@ -2,7 +2,7 @@
...
+The PricingQuantity column MUST be present in the billing data. This column MUST be of type Decimal and MUST conform to [Numeric Format](#numericformat) requirements. The value MAY be negative in cases where [ChargeClass](#chargeclass) is "Correction". This column MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric, except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```
Comment (flanakin) [inline specification/columns/pricingquantity.md:5]:
```suggestion
The PricingQuantity column MUST be present in the billing data. This column MUST be of type Decimal and MUST conform to [Numeric Format](#numericformat) requirements. The value MAY be negative in cases where [ChargeClass](#chargeclass) is "Correction". This column MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric, except in cases of ChargeClass "Correction", which MAY address PricingQuantity or any cost discrepancies independently.
```

**Comment by ijurica** [comment]
Suggested changes applied (changed 'MAY' to 'may'; replaced 'and/or' with 'or'; updated Changelog)
Ready for final review, approval, and hopefully merging.

**Comment by jpradocueva** [comment]
@mike-finopsorg, @udam-f2 can one of you please approve this PR? It is ready to merge.

**Review by mike-finopsorg — APPROVED** [review]

**Pull request merged**